### ISSUE TITLE
Dune should run itself when a package build command tries to run `dune`

### ISSUE THREAD

**Issue by gridbugs** [ocaml/dune#10019]
If the dune executable currently running isn't what `dune` resolves to in `PATH`, packages that invoke `dune` in their build command should still start a copy of the current dune rather than the one from `PATH`. This situation arises when running a development version of dune where the exe is named `main.exe`.

**Comment by rgrinberg** [comment]
note that this should extend to `(run sh -c dune)` as well. In other words, we should set `PATH` so that the current dune is included.

**Comment by gridbugs** [comment]
Good point; it's not sufficient to detect when the first word of the command is `dune` and replace it with the value of `argv[0]`. Setting `PATH` is also insufficient as it wouldn't handle the case where your dune binary is named something else (e.g. if you build dune yourself and it's still in `_build/default/bin/main.exe`). What about creating a temporary directory somewhere and copying/linking the current dune exe there (named "dune") and then putting that directory at the start of `PATH`? Is that too unix-specific for it to work on windows?

**Comment by dra27** [comment]
@gridbugs - that trick is entirely fine for Windows and indeed opam’s prototype compiler packaging is using a very similar trick to shim GCC out of Cygwin

**Comment by Leonidas-from-XIV** [comment]
You can get the current executable also on Windows [like this](https://github.com/dra27/ocaml/blob/enable-relative/runtime/win32.c#L671-L687) (from @dra27's relocatable compiler branch), however unlike `Unix.readlink` on `/proc/self/exe` probably not in pure OCaml.

**Comment by kit-ty-kate** [comment]
I'm not sure to understand what is the advantage of using dune recursively? How does this work if the constraint for dune doesn't match the current one? Isn't that also contrary to the idea of a lock file?

**Comment by gridbugs** [comment]
@kit-ty-kate good point. And the lockdir does contain a lockfile for dune when a project's dependencies depend on dune, so we should be building dune as part of the project's build and making the resulting exe available to all packages that depend on it. I wonder that's not currently happening in my experiments. I'll investigate further.

**Comment by rgrinberg** [comment]
>  What about creating a temporary directory somewhere and copying/linking the current dune exe there (named "dune") and then putting that directory at the start of PATH? Is that too unix-specific for it to work on windows?

This is likely the best.

>  so we should be building dune as part of the project's build and making the resulting exe available to all packages that depend on it

That can't work because the dune that will be build will read/write metadata in a different format.

**Comment by gridbugs** [comment]
> That can't work because the dune that will be build will read/write metadata in a different format.

Which metadata files are you referring to? If it's just metadata files within the package it's building then what would be the problem?

**Comment by rgrinberg** [comment]
The `dune-package` fill produced when building a package must be compatible. This is what dune reads to make libraries, package sites, and other metadata available.

**Comment by gridbugs** [comment]
But why does the dune that's orchestrating the build need to read the `dune-package` files of individual packages from the project's dependencies? Dependencies don't need to build with dune at all so I don't see why dune can't treat the dune package as a foreign build system that happens to be called "dune".

**Comment by rgrinberg** [comment]
Which dune do you plan to use to build the packages that are inside the user's workspace? E.g.

```
$ dune build bin/foo.exe
```

Will that dune used to install dune, or the dune that the user already had?

**Comment by gridbugs** [comment]
For packages in the current workspace I think it makes sense to use the dune the user already has (ie. the one orchestrating the build). But for packages in the lockdir that build with dune, I'm considering treating their `dune` dependency the same way we could treat any other build dependency of a package, building the dune package of the appropriate version and making its binary available to build the package.

**Comment by rgrinberg** [comment]
> For packages in the current workspace I think it makes sense to use the dune the user already has

Okay, so this dune will not be able to read the metadata generated by the inner dune. To discover the metadata about the packages we just built, we need to read the `dune-package` files.

**Comment by gridbugs** [comment]
Why is it important that the outer dune be able to read metadata generated by the inner dune? Not all packages build with dune and the outer dune can still handle the ones that don't.

**Comment by rgrinberg** [comment]
For packages that are built with dune, we rely on findlib to provide partial metadata. It's important to note that this metadata is limited, and dune packages can provide much richer metadata which we do indeed use to provide many advanced dune features.

**Comment by gridbugs** [comment]
Wait did you mean to say that packages that _aren't_ built with dune rely on findlib to provide partial metadata? I imagine we wouldn't need findlib when working with packages built with the same version of dune.

**Comment by rgrinberg** [comment]
Yes, my bad. The findlib metadata is needed for non dune packages only.

**Comment by gridbugs** [comment]
Makes sense. Can you elaborate on the advanced dune features that will take advantage of cases where non-workspace dependencies build with dune (or point to documentation). Also it sounds like we should constrain dependencies to those compatible with the current version of dune. Is that right?

**Comment by rgrinberg** [comment]
> Can you elaborate on the advanced dune features that will take advantage of cases where non-workspace dependencies build with dune (or point to documentation)

Here's a partial list: ppx rewriters, ppx drivers, virtual libraries, inline test drivers, privately installed libraries, custom installation sites. Essentially, anytime we need to implement some advanced features, we need to tweak this metadata to make it work for installed dune packages. This metadata has no compatibility guarantees and we are able to modify it whenever we want.

>  Also it sounds like we should constrain dependencies to those compatible with the current version of dune. Is that right?

We already do that. Sounds like we should drop dune from the lock directory.

By the way, our end goal is not to launch any external dune processes to build packages. This is going to give us some rather significant performance improvements.

**Comment by shonfeder** [comment]
> our end goal is not to launch any external dune processes to build packages. This is going to give us some rather significant performance improvements.

This sounds nice to me. But it also counter to what is proposed and discussed here, afaict. Where is this end goal currently tracked? I'd like to understand better how we will get there, since it seems to be a prerequisite for #12103.

**Comment by shonfeder** [comment]
Including this in https://github.com/ocaml/dune/milestone/62, as suggested by @Alizter, since we believe it is a prerequisite for fixing https://github.com/ocaml/dune/issues/12103

**Comment by shonfeder** [comment]
Removing this from the MSP for the reason noted on https://github.com/ocaml/dune/issues/12103#issuecomment-3500660338

**Comment by Leonidas-from-XIV** [comment]
A test reproducing the incorrect behavior was implemented and merged in #10020 (and then duplicate repros were written in #12698 and #12845).

**Comment by Leonidas-from-XIV** [comment]
I've created two PRs to fix the issue:

 * [x] #12902 changes the resolving of binaries of `(run)` actions in package rules. When evaluating run rules Dune first resolves the binary that it should invoke. This PR special-cases that if the binary is `dune` then it always resolves to its own binary. Thus `(run dune build)` will run the correct dune.
 * [x] #12935 extends the `PATH` of the executables to be run to prepend a directory containing the currently running dune. Thus `(run sh -c 'dune')` will launch the currently running dune.

Together these PRs fix the 3 repro cases.
